### PR TITLE
Multithread gradient descent and compute cost every 50 iterations

### PR DESCRIPTION
**Pull request by linqiaozhi**
Two improvements:

1. The computation of the attractive term at each iteration now is now multithreaded, which gives approximately 3x speed-up for datasets of ~1 million points on multithreaded machines. 
2. The approximate KL divergence is computed every 50 iterations and output to the console. R wrapper has been updated to read this in.

### CODE REVIEW

**Comment by dkobak** [comment]
Sounds good. Does 3x speed up refer to the computation of the attractive term or to the the overall gradient descent time? I have no idea what fraction of time is taken by the attractive vs. repulsive computations.

**Comment by linqiaozhi** [comment]
The computation of the repulsive force was essentially rendered negligible by our interpolation scheme, so the bottleneck (with large numbers of points) was simply looping through the perplexity*3 number of neighbors to compute the attractive force at each iteration. This new commit parallelizes that loop, with some pretty nice results on multi-core machines.

For example, on my Macbook Pro, Barnes-Hut takes 1 hour and 10 min for 1000 iterations of gradient descent with 500,000 points. FIt-SNE now takes 3 minutes.  

Or on our server, which has 24 (slower) cores, it Barnes Hut takes almost 9 hours for 1000 iterations of 1 million points. FIt-SNE takes 11 min. That's about 50 times faster. 

In other words, with a single core, we are 6-10 times faster than Barnes Hut. With multiple cores and this new commit, we can be up to 50 times faster. 

Note that this multithreading is only this useful because of the interpolation scheme, because without it, the computation of repulsion term dominates.

Would be happy to hear if you see similar boosts in performance on your hardware (whenever you have a chance to try)! If you pull the most recent version, be sure to compile though!

**Comment by dkobak** [comment]
Oh I see. I did not realize that it's the attractive force computations that were such a bottleneck!

The numbers are very impressive. Can't wait to try it out when I get back to the lab :)

**Comment by dkobak** [comment]
I quickly tried it now with n=250k. The version of the code before this improvement took 20s/50 iterations on my lab desktop (4 cores with 2 threads/core, not sure if it counts as 4 or as 8), the current version takes 7 seconds. So that's 3x improvement.

For 1.3 mln points I think it used to take ~20-25 min / 50 iterations (don't remember exactly and don't want to rerun it now to check), the current version takes 120sec, so this looks like 10x speed-up (!). Really impressive.

I am wondering if one can speed it up even further by moving the computation of attractive forces to GPU. Have you seen this: https://arxiv.org/pdf/1807.11824.pdf ? Coincidentally I came across it earlier today. They seem to use Barnes-Hut for repulsion but use CUDA for attractive forces. They report 50-100x improvement compared to BH, when using a powerful GPU. I guess yours can be even faster. I have zero experience with CUDA in C++ so I don't know how difficult it is to let a multithreaded loop run on GPU... Looking at their source code https://github.com/CannyLab/tsne-cuda, it seems they rewrote the whole thing.

**Comment by dkobak** [comment]
Oh, interesting! So when you said this: "The computation of the repulsive force was essentially rendered negligible by our interpolation scheme", it was a bit of an overstatement? :-)

**Comment by linqiaozhi** [comment]
@dkobak Haha it was true before!! As in, compared to the single threaded computation of the attractive force, the repulsive force was negligible for large N. Now that the computation of the attractive force is multithreaded, the bottleneck is again the repulsive force, so it's worth it to multithread it :)
